src/main/java/hudson/matrix/MatrixBuild.java

```diff
@@ -91,34 +91,32 @@ public Object readResolve() {
     }
 
     /**
+     * Delete build and its child builds.
+     *
+     * Child builds can be skipped if they have to be kept.
+     *
+     * @since TODO
+     */
+    @Override
+    public void delete() throws IOException {
+        super.delete();
+
+        for(MatrixRun run : getExactRuns()) {
+            if (run.getWhyKeepLog() != null) continue;
```

Comment:
I would expect that `MatrixBuild.getWhyKeepLog` would return non-null in this case, so that the whole thing is kept as a unit.